### PR TITLE
Heuristics to minimize performance impact of `-XX:+DebugOnRestore`

### PR DESCRIPTION
**Pull request by dsouzai**
This PR includes several heuristics to minimize the performance impact of `-XX:+DebugOnRestore`:
- Generate FSD compilations pre-checkpoint
- Compile "important" methods pre-checkpoint
  - "important" methods are those that are in the SCC, that would have
    been loaded at scount=20
  - These "important" methods are compiled without FSD enabled
- Compile failed compilations pre-checkpoint as opposed to post-restore
- Reset environment after proactive compilation 
- Don't change JVM Phase pre-checkpoint
- Enable IProfiling during startup; because SCC validation occurs post
  restore, there is no persistent IProfiling data to pull from
- Turn off IProfiling on restore so that the normal heuristics can turn
  it back on post-startup
- Trigger recomps pre-checkpoint 

Parent issue https://github.com/eclipse-openj9/openj9/issues/18866

### CODE REVIEW

**Comment by dsouzai** [comment]
Hm I thought I covered that scenario, but it looks like I only partially addressed it, namely that if FSD is specified post-restore then FSD remains enabled. But, there needs to be a way to invalidated the methods that were compiled proactively. 

I guess originally the idea was to "commit" the startPCs of these proactive methods on restore, in which case if someone specified FSD post-restore then we could just skip the "commit". However, since "commit"-ing post-restore hurts footprint, I updated the `extra` field pre-checkpoint, but then now those methods will not support FSD.

A relatively simple solution to this would be to just invalidate the entire code cache if someone specifies FSD post-restore. All the stacks at that point will be at a safe point and they will be running FSD code, so even though the extra field would be updated, it wouldn't matter from a OSR point of view.

Otherwise, I'll need another list to keep track of methods that were compiled proactively and revert the extra field of those methods if FSD is specified post-restore. Maybe that's something I can do in another PR since it's fairly non-trivial, and this PR is already reasonably complicated.

I'll make the simple change to this PR tomorrow.

**Comment by dsouzai** [comment]
@mpirvu good for review again.

File: runtime/compiler/runtime/CRRuntime.cpp [hunk 1]
```diff
@@ -437,6 +403,50 @@ bool TR::CRRuntime::compileMethodsForCheckpoint(J9VMThread *vmThread)
       return false;
       }
 
+      {
+      J9JavaVM *javaVM = getJITConfig()->javaVM;
+      TR_J9VMBase *fej9 = TR_J9VMBase::get(getJITConfig(), vmThread);
+
+      TR::CRRuntime::SetupEnvForProactiveComp setupProactiveCompEnv(this, javaVM, vmThread, fej9);
```
Comment (mpirvu) [inline runtime/compiler/runtime/CRRuntime.cpp:410]: Do we need `TR::CRRuntime::` here?

**Review by mpirvu — COMMENTED** [review]
If FSD is enabled post-restore, the entire code cache is invalidated. Then why bother to generate warm FSD bodies pre-checkpoint?

File: runtime/compiler/control/CompilationThread.cpp [hunk 2]
```diff
@@ -8545,7 +8547,18 @@ TR::CompilationInfoPerThreadBase::wrappedCompile(J9PortLibrary *portLib, void *
                   aotCompilationReUpgradedToWarm = true;
                   }
                }
-
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+            else if (vmThread->javaVM->internalVMFunctions->isDebugOnRestoreEnabled(vmThread)
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/control/CompilationThread.cpp:8551]: Yeah I could, it's been a while but I think the reason I wanted to put it here was to ensure that the opt level was never cold (unless I guess specified via -Xjit). If having this code in `shouldDowngradeToCold` will ensure that then I can move it there.

**Comment by dsouzai** [comment]
> If FSD is enabled post-restore, the entire code cache is invalidated. Then why bother to generate warm FSD bodies pre-checkpoint?

I guess because originally we were going to do the recompilation post-restore, so it was going to be easier to just continue using the FSD code generated pre-checkpoint. However, I guess at least in this PR, I'm just invalidating the code cache as a first simple solution; maybe if we realize that it's too hard to keep track of all the methods that need to be reset back to FSD, then there's no point having warm FSD (though there may be some performance impact due to a difference in iprofiling data and what not).

File: runtime/compiler/control/CompilationThread.cpp [hunk 3]
```diff
@@ -660,6 +660,19 @@ bool TR::CompilationInfo::shouldDowngradeCompReq(TR_MethodToBeCompiled *entry)
       TR_J9VMBase *fe = TR_J9VMBase::get(_jitConfig, NULL);
       const J9ROMMethod * romMethod = methodDetails.getRomMethod(fe);
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+      J9VMThread *vmThread = _jitConfig->javaVM->internalVMFunctions->currentVMThread(_jitConfig->javaVM);
+      if (vmThread->javaVM->internalVMFunctions->isDebugOnRestoreEnabled(vmThread)
+          && vmThread->javaVM->internalVMFunctions->isCheckpointAllowed(vmThread)
+          && !entry->_checkpointInProgress
+          && entry->_optimizationPlan->isOptLevelDowngraded()
```
Comment (mpirvu) [inline runtime/compiler/control/CompilationThread.cpp:668]: We don't need `entry->_optimizationPlan->isOptLevelDowngraded()` and `entry->_optimizationPlan->getOptLevel() == cold` tests here. They will never be true.

**Comment by dsouzai** [comment]
@mpirvu should be good for review now.

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by mpirvu** [comment]
jenkins test sanity xlinux,zlinux jdk21

**Comment by mpirvu** [comment]
jenkins compile win jdk17

**Comment by mpirvu** [comment]
Compilation errors on Windows:
```
f:\users\jenkins\workspace\build_jdk17_x86-64_windows_personal\openj9\runtime\compiler\control\CompilationThread.cpp(4173): error C2039: 'getCRRuntime': is not a member of 'TR::CompilationInfo'
12:02:11  f:\users\jenkins\workspace\build_jdk17_x86-64_windows_personal\openj9\runtime\compiler\runtime/IProfiler.hpp(76): note: see declaration of 'TR::CompilationInfo'
12:02:11  [ 41%] Building C object runtime/jcl/CMakeFiles/jclse.dir/common/mgmtcompilation.c.obj
12:02:11  RegionValidator.cpp
12:02:11  make[6]: *** [runtime/compiler/CMakeFiles/j9jit.dir/build.make:371: runtime/compiler/CMakeFiles/j9jit.dir/control/CompilationThread.cpp.obj] Error 2
```

**Comment by dsouzai** [comment]
jenkins compile win jdk17

**Pull request merged**